{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Exam 2 Answers* / Author: **Lucia Sanchez Gonzalez**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Germplasm = open(\"Germplasm.tsv\", \"r\") #variable that contains the content of the file \"Germplasm\" by opening and reading it with the function open\n",
    "LocusGene = open(\"LocusGene.tsv\",\"r\")# variable that contains the content of the file \"LocusGene\" \n",
    "\n",
    "def obtain_col(file,number): # a function that goes through the file we give as an argument, extracts the data of the column we indicate and adds them to a list\n",
    "    col_list = []\n",
    "    for line in file.readlines()[1:]: #a loop that goes through every line (except the header) and adds to the list the colum of the file that we indicate\n",
    "       col_list = col_list + [line.split(\"\\t\")[number]]\n",
    "    return(col_list)\n",
    "\n",
    "AGI_germplasm = obtain_col(Germplasm,number = 0)#we apply the function giving as arguments the file and the column (in this case the Locus) that we want\n",
    "AGI_LocusGene = obtain_col(LocusGene,number = 0)\n",
    "\n",
    "for AGI_G in AGI_germplasm: #a loop that goes through every AGI locus in the Germplasm file\n",
    "    if AGI_G in AGI_LocusGene: #if that AGI locus is also in the LocusGene file, the data of both is in the same sequence\n",
    "        print(AGI_G + \" is in both files\")\n",
    "    else: #if there is an AGI locus of Germplasm file that isn't in the LocusGene file it will print the next:\n",
    "        print(\"Not all the lines of data in Germplasm.tsv and LocusGene are in the same sequence :-(\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this exercise, I decided to create a function that allows us to get the data of the column I give as an argument of the file (also as an argument). This function allows us to get get a list with the AGI Locus of both files without repeat the same code. Then, I used a loop \"for\" and a condition \"if\" to compare the AGI Locus of both files; with this type of loop, I doesn't matter if the locus are in different orders in the lists. Because there isn't too many Locus (rows), I decided to print the name of each to indicate that it is in both file, but if there were much more rows, I would probably type the next commands:\n",
    "\n",
    "\n",
    "for AGI_G in AGI_germplasm: \n",
    "\n",
    "    if AGI_G in AGI_LocusGene: \n",
    "    \n",
    "        print(\"\")\n",
    "        \n",
    "    else: \n",
    "    \n",
    "        print(\"Not all the lines of data in Germplasm.tsv and LocusGene are in the same sequence :-(\")\n",
    "        \n",
    "\n",
    "and then if there were something wrong, we would only see the message in the \"else\" line. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this problem I decided to use sqlMagic. First, I have to type this in my terminal:\n",
    "\n",
    "     docker start course-mysql \n",
    "\n",
    "and then:\n",
    "\n",
    "    mysql -h 127.0.0.1 -P 3306 --protocol=tcp -u root -p\n",
    "    \n",
    "This allows us to connect to MySQL in Docker."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#To use sqlMagic first we have to run the next commands:\n",
    "\n",
    "%load_ext sql #we starts sql to conecte and modify our database. Since we'd already load it,we just have to \"reload\"\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "\n",
    "## To create a new database and use it, we run these lines:\n",
    "%sql create database mydatabase; #creates a database call \"mydatabase\" \n",
    "%sql use mydatabase; #tells the server that we want to interact only with our database we just created\n",
    "\n",
    "### Creates a table call \"germplasm\" that contains 5 columns ##\n",
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(20)NOT NULL, Germplasm VARCHAR(200) NOT NULL, phenotype VARCHAR(600)NOT NULL, pubmed INTEGER NOT NULL);\n",
    "\n",
    "### Creates a table call \"LocusGene\" that contains 4 columns ##\n",
    "%sql CREATE TABLE LocusGene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(20) NOT NULL, Gene VARCHAR(15) NOT NULL, Protein_Length INTEGER NOT NULL);\n",
    "\n",
    "%sql SELECT * FROM germplasm INNER JOIN LocusGene ON germplasm.id = LocusGene.id; #It shows the two tables joined by the id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I created a database call \"mydatabase\" and then I indicated to the server that I want to use it. Then I created  two tables:\n",
    "\n",
    "For example, table germplasm contains five columns:\n",
    "\n",
    "*id: It will be the key that will allows us to connect both tables. We indicate that it is a number with \"INTEGER\"; it will increment as the number of rows grow with \"AUTO_INCREMENT\", and that it will be the key with \"PRIMARY KEY\".\n",
    "\n",
    "*Locus: We indicate that it is a string of characters (numbers, letters, etc) and that it will a maxime length with \"VARCHAR(number)\". With NOT NULL, we indicate that there cannot be a blank space in this column.\n",
    "\n",
    "The rest of its columns and the table LocusGene (with 4 columns) also use these kind of arguments.\n",
    "\n",
    "Then, to show that these two tables are linked in a 1:1 relationship, I use the command:\n",
    "\n",
    "    %sql SELECT * FROM germplasm INNER JOIN LocusGene ON germplasm.id = LocusGene.id;\n",
    "\n",
    "This allows us to see the two table joined by the id column, and we can see that both have the same number of rows.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with.\n",
    "\n",
    "\n",
    "For this exercise I decided to use the csv library, because It allows us to obtain every row of each file as a dictionary (with the function \"csv.DictReader\") where the keys are the columns name we gave with fieldnames. Even though this library is normally used with csv, here we can use it with tsv file with the argument \"delimiter = \"\\t\" in the csv.DictReader function. csv.DictReader gives us a list that contains every row of the file as a dictionary, so with a loop \"for\" I can go through every row and with the command INSERT INTO table (colum1 name,colum2 name) VALUES(row[\"name of the column\"],...) we can insert the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import csv  # import the csv library\n",
    "import pymysql.cursors #import the object cursors from the pymsql library\n",
    "connection = pymysql.connect(host='localhost',  #we stablish connection with our database\n",
    "                             user='root', \n",
    "                             password='root',\n",
    "                             db='mydatabase', #database we want to use\n",
    "                             charset='utf8mb4', \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "\n",
    "try: \n",
    "    with connection.cursor() as cursor:\n",
    "       \n",
    "        with open('Germplasm.tsv') as csvfile: #creates a dictionary that contains the data of the file; its keys are the columns\n",
    "            data_germplasm = csv.DictReader(csvfile, fieldnames=(\"Locus\",\"germplasm\",\"phenotype\",\"pubmed\"),delimiter='\\t') #to read the file we have to indicate that the data is delimiter by \"\\t\"\n",
    "            next(data_germplasm,None) #to skip the header with the column' names\n",
    "            for row in data_germplasm: #for every row (that is a dictionary) of the file we read\n",
    "                    sql = \"\"\"INSERT INTO germplasm (Locus,Germplasm,phenotype,pubmed)\n",
    "                    VALUES('\"\"\"+row['Locus']+\"\"\"', '\"\"\"+row['germplasm']+\"\"\"','\"\"\"+row['phenotype']+\"\"\"',\"\"\"+row['pubmed']+\"\"\")\"\"\" #insert in the table each data of each row\n",
    "                    cursor.execute(sql) #execute the sql query \n",
    "        \n",
    "        with open('LocusGene.tsv') as csvfile: #same procedure for the second table\n",
    "            data_locusgene = csv.DictReader(csvfile, fieldnames=(\"Locus\",\"Gene\",\"ProteinLength\"),delimiter='\\t',quotechar='\"') \n",
    "            next(data_locusgene,None) #to skip the header with the colums' names\n",
    "            for row in data_locusgene: #for every row (that is a dictionary) of the file we read\n",
    "                sql = \"\"\"INSERT INTO LocusGene (Locus,Gene,Protein_Length)\n",
    "                VALUES('\"\"\"+row['Locus']+\"\"\"', '\"\"\"+row['Gene']+\"\"\"',\"\"\"+ row['ProteinLength']+\"\"\")\"\"\" #insert in the table each data of each row\n",
    "                cursor.execute(sql) #execute the sql query \n",
    "    connection.commit() #commits the changes to the database.\n",
    "finally:\n",
    "    connection.close() #close the connections with the database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish.\n",
    "\n",
    "\n",
    "\n",
    "* For ***Report 1*** I created a function that joins two tables by two columns we indicate as arguments (we also have to indicate the database we want to use). It returns us a list that contains each row as a dictionary. I used that function to obtain a list with the rows of the joined table. Then I created a loop for with the keys of the first row (the first element of the list, a dictionary) to obtain the header. Then I used a similar loop to create a string with the data.\n",
    "\n",
    "* For ***Report 2*** I created a similar function that I used in report 1 but in with this function we can indicate which rows of the tables we want to joined by indicating a data in a particular column. In this case, I used it to joined the rows of the table that contain \"SKOR\" and \"MAA3\" in the column \"Gene\". Then I use the same procedure that in the report 1 to obtain a string with the data. \n",
    "\n",
    "* For ***Report 3 and 4*** I created a function that allows us to obtain certain data from the table we want bu using a regular expression. In report 3 I used it to count the entries of each chromosome giving in the argument \"function\": COUNT(*). With the regular expression and the name of the column it searchs for all the entries and then counts how many are. In report 4 I used the same function but now I indicate that I want the average of the protein length of the entries of each chromosome with this argument: AVG(Protein_Length)\". Then I saved the results in a variable that contains the data as strings.\n",
    "\n",
    "\n",
    "Finally to save all reports in just one file, I use first the functions open(\"name_file\", \"r\") and name_file.write() to start the new file and start to write in it. Then, to add the rest of reports, I used the argument \"a\" in the function open() to \"append\" the new data in the file without erasing the ones that already has."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### 1. A report that shows the full, joined, content of the two database tables (including a header line) ######\n",
    "\n",
    "def jointables(database,table1, table2,col1,col2): #a function that joins two tables by the columns we indicate as arguments\n",
    "    import pymysql.cursors #import the object cursors from the pymsql library\n",
    "    connection = pymysql.connect(host='localhost',  #we stablish connection with our database\n",
    "                             user='root', \n",
    "                             password='root',\n",
    "                             db= database, #database we want to use\n",
    "                             charset='utf8mb4', \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM {} INNER JOIN {} WHERE \n",
    "     {} = {};\"\"\".format(table1,table2,col1,col2) #.format allows us to insert the data (in this case as arguments of the function) in the order that we want \n",
    "        cursor.execute(sql)\n",
    "        result = cursor.fetchall() #save in this variable the results of the command we execute before\n",
    "        return(result)\n",
    "    \n",
    "results_1 = jointables(\"mydatabase\",\"germplasm\",\"LocusGene\",\"germplasm.id\",\"LocusGene.id\") # this is a list with every row resulting of join the two tables\n",
    "\n",
    "\n",
    "report1_header = \"\" #empty string that will contain the header\n",
    "for row in results_1[0].keys(): #a loops that allows to insert the column' names in the string to use it as the header\n",
    "    report1_header = report1_header + \"\\t\" + row\n",
    "\n",
    "report1_data = \"\" #empty string that will contain the report 1\n",
    "for row_data in results_1: #a loops that allows to insert the column' data in the string where every data is separated by a \"\\t\"\n",
    "    report1_data += str(row_data[\"id\"])+ \"\\t\" + row_data[\"Locus\"]+ \"\\t\"+ row_data[\"Germplasm\"]+ \"\\t\"+ row_data[\"phenotype\"]+ \"\\t\" \\\n",
    "    +str(row_data[\"pubmed\"])+ \"\\t\"+ str(row_data[\"LocusGene.id\"])+ \"\\t\"+ row_data[\"LocusGene.Locus\"]+ \"\\t\"+ row_data[\"Gene\"]\\\n",
    "    + str(row_data[\"Protein_Length\"])+ \"\\n\"\n",
    "\n",
    "    \n",
    "########### 2. Create a joined report that only includes the Genes SKOR and MAA3 ###################################\n",
    "\n",
    "def jointables_2(database,table1, table2,col1,col2,row,gene): #a function that joins two tables by the columns and rows we indicate as arguments\n",
    "    import pymysql.cursors\n",
    "    connection = pymysql.connect(host='localhost',  \n",
    "                             user='root', \n",
    "                             password='root',\n",
    "                             db= database, #database we want to use\n",
    "                             charset='utf8mb4', \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM {} INNER JOIN {} WHERE \n",
    "     {} = {} AND {} LIKE \"{}\";\"\"\".format(table1,table2,col1,col2,row,gene)#give us a table where are joined two rows depending of the data of the column we indicate\n",
    "        cursor.execute(sql)\n",
    "        result = cursor.fetchall()\n",
    "        return(result)\n",
    "    \n",
    "results_2_SKOR = jointables_2(\"mydatabase\",\"germplasm\",\"LocusGene\",\"germplasm.id\",\"LocusGene.id\",\"Gene\",\"SKOR\")#obtain a list that contains the row of Gene SKOR as a dictionary \n",
    "results_2_MAA3 = jointables_2(\"mydatabase\",\"germplasm\",\"LocusGene\",\"germplasm.id\",\"LocusGene.id\",\"Gene\",\"MAA3\")#obtain a list that contains the row of Gene MAA3 as a dictionary\n",
    "results_2 =  results_2_SKOR + results_2_MAA3 #join in one list the rows (dictionaries) of both genes\n",
    "\n",
    "report2_data = \"\" \n",
    "for row_data in resultados_2:\n",
    "    report2_data += str(row_data[\"id\"])+ \"\\t\" + row_data[\"Locus\"]+ \"\\t\"+ row_data[\"Germplasm\"]+ \"\\t\"+ row_data[\"phenotype\"]+ \"\\t\" \\\n",
    "    +str(row_data[\"pubmed\"])+ \"\\t\"+ str(row_data[\"LocusGene.id\"])+ \"\\t\"+ row_data[\"LocusGene.Locus\"]+ \"\\t\"+ row_data[\"Gene\"]\\\n",
    "    + str(row_data[\"Protein_Length\"])+ \"\\n\"\n",
    "\n",
    "    \n",
    "######### 3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx) #########\n",
    "\n",
    "def operations_entries(database,function,table,col,regexp): #function that allows use to get data from the table by using a regular expression\n",
    "    import pymysql.cursors\n",
    "    connection = pymysql.connect(host='localhost',  \n",
    "                             user='root', \n",
    "                             password='root',\n",
    "                             db= database, #database we want to use\n",
    "                             charset='utf8mb4', \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT {} FROM {} WHERE {}\n",
    "     REGEXP \"{}\";\"\"\".format(function,table,col,regexp)#we must give a function that we want to use with the data, the table name, the column where is the data, and the regula expression\n",
    "        cursor.execute(sql)\n",
    "        result = cursor.fetchall()#it gives us a list that has the next kind of element: a dictionary that contain the data we want\n",
    "        return(result)\n",
    "\n",
    "num_entreis_chr1 = operations_entries(\"mydatabase\",\"COUNT(*)\",\"germplasm\",\"Locus\",\"AT1G\")#use the function to count how many entries of chromosome 1 are.\n",
    "num_entreis_chr2 = operations_entries(\"mydatabase\",\"COUNT(*)\",\"germplasm\",\"Locus\",\"AT2G\")#use the function to count how many entries of chromosome 2 are.\n",
    "num_entreis_chr3 = operations_entries(\"mydatabase\",\"COUNT(*)\",\"germplasm\",\"Locus\",\"AT3G\")#use the function to count how many entries of chromosome 3 are.\n",
    "num_entreis_chr4 = operations_entries(\"mydatabase\",\"COUNT(*)\",\"germplasm\",\"Locus\",\"AT4G\")#use the function to count how many entries of chromosome 4 are.\n",
    "num_entreis_chr5 = operations_entries(\"mydatabase\",\"COUNT(*)\",\"germplasm\",\"Locus\",\"AT5G\")#use the function to count how many entries of chromosome 5 are.\n",
    "num_total_chr = num_entreis_chr1 + num_entreis_chr2+num_entreis_chr3+num_entreis_chr4+num_entreis_chr5 #concatenate the lists with the numbers \n",
    "\n",
    "report3_data = \"\" #empty list\n",
    "chrm = 0 #it will be used to indicate the number of chromosome\n",
    "for num in num_total_chr: #for every element of the list with the number of entries of each chromosomes:\n",
    "    chrm += 1 #sums one to the value of the variable\n",
    "    report3_data += \"The chromosome \"+ str(chrm) + \" has: \" + str(num[\"COUNT(*)\"]) + \" entries\" + \"\\n\" #prints the number of entries for each chromosome\n",
    "\n",
    "\n",
    "######## 4. Create a report that shows the average protein length for the genes on each Chromosome ##########\n",
    "\n",
    "averg_chr_1 = operations_entries(\"mydatabase\",\"AVG(Protein_Length)\",\"LocusGene\",\"Locus\",\"AT1G\")#use the function to get the average of protein length of the entries of chromosome 1.\n",
    "averg_chr_2 = operations_entries(\"mydatabase\",\"AVG(Protein_Length)\",\"LocusGene\",\"Locus\",\"AT2G\")#use the function to get the average of protein length of the entries of chromosome 2.\n",
    "averg_chr_3 = operations_entries(\"mydatabase\",\"AVG(Protein_Length)\",\"LocusGene\",\"Locus\",\"AT3G\")#use the function to get the average of protein length of the entries of chromosome 3.\n",
    "averg_chr_4 = operations_entries(\"mydatabase\",\"AVG(Protein_Length)\",\"LocusGene\",\"Locus\",\"AT4G\")#use the function to get the average of protein length of the entries of chromosome 4.\n",
    "averg_chr_5 = operations_entries(\"mydatabase\",\"AVG(Protein_Length)\",\"LocusGene\",\"Locus\",\"AT5G\")#use the function to get the average of protein length of the entries of chromosome 5.\n",
    "averg_total_chr = averg_chr_1 + averg_chr_2 + averg_chr_3 + averg_chr_4 + averg_chr_5 #Concatenate the lists with the number of average length.\n",
    "\n",
    "report4_data = \"\"#empty list\n",
    "chrm = 0 #it will be used to indicate the number of chromosome\n",
    "for number in averg_total_chr:\n",
    "    chrm += 1\n",
    "    report4_data += \"The average protein length for the genes of the chromosome \"+ str(chrm) + \" is: \" + str(number[\"AVG(Protein_Length)\"]) + \"\\n\"\n",
    "\n",
    "\n",
    "\n",
    "######################## SAVE ALL THE REPORTS IN A FILE ###################################\n",
    "\n",
    "\n",
    "final_file = open(\"Reports_problem_4.tsv\",\"w\") #creates a new file called \"Reports_problem_4.tsv\"\n",
    "final_file.write(\"REPORT 1: \" + \"\\n\") #adds that string into the file\n",
    "final_file.close() #close the file \n",
    "\n",
    "final_file = open(\"Reports_problem_4.tsv\",\"a\") #open the file to add more information without removing the original content\n",
    "final_file.write(report1_header + \"\\n\") #appends the report 1 header (the column' names) and a new line\n",
    "final_file.write(report1_data + \"\\n\") #appends the report 1 and a new line\n",
    "final_file.write(\"REPORT 2: \" + \"\\n\")\n",
    "final_file.write(report1_header + \"\\n\")\n",
    "final_file.write(report2_data + \"\\n\")\n",
    "final_file.write(\"REPORT 3: \" + \"\\n\")\n",
    "final_file.write(report3_data + \"\\n\")\n",
    "final_file.write(\"REPORT 4: \" + \"\\n\")\n",
    "final_file.write(report4_data + \"\\n\")\n",
    "final_file.close() #close the file to stop editing \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
